addons/purchase/tests/test_purchase_downpayment.py

```diff
@@ -0,0 +1,686 @@
+        dp_invoice.invoice_date = datetime.datetime.today()
+        dp_invoice.action_post()
+
+        self.assertRecordValues(pol_downpayment, [{
+            'price_unit': 587.5,
+            'discount': 0.0,
+            'product_qty': 0.0,
+            'qty_invoiced': 1.0,
+            'qty_to_invoice': -1.0,
+        }])
+        self.assertEqual(self.purchase_order.invoice_status, 'to invoice')
+
+        self.make_downpayment(advance_payment_method='delivered')
+
+        po_invoice = max(self.purchase_order.invoice_ids)
+        self.assertEqual(len(po_invoice.invoice_line_ids.filtered(lambda l: not (l.display_type == 'line_section' and l.name == "Down Payments"))),
+                         len(self.purchase_order.order_line.filtered(lambda l: not (l.display_type == 'line_section' and l.name == "Down Payments"))), 'All lines should be invoiced')
+        self.assertEqual(len(po_invoice.invoice_line_ids.filtered(lambda l: l.display_type == 'line_section' and l.name == "Down Payments")), 1, 'A single section for downpayments should be present')
+        self.assertEqual(po_invoice.amount_total, self.purchase_order.amount_total - pol_downpayment.price_unit, 'Downpayment should be applied')
+        po_invoice.invoice_date = datetime.datetime.today()
+        po_invoice.action_post()
+
+        credit_note_wizard = self.env['account.move.reversal'].with_context({'active_ids': [po_invoice.id], 'active_id': po_invoice.id, 'active_model': 'account.move'}).create({
+            'reason': 'reason test refund with downpayment',
+            'journal_id': po_invoice.journal_id.id,
+        })
+        credit_note_wizard.refund_moves()
+        invoice_refund = self.purchase_order.invoice_ids.sorted(key=lambda inv: inv.id, reverse=False)[-1]
+        invoice_refund.invoice_date = datetime.datetime.today()
+        invoice_refund.action_post()
+
+        self.assertEqual(self.pol_product_order.qty_to_invoice, 5.0, "As the refund still exists, the quantity to invoice is the ordered quantity")
+        self.assertEqual(self.pol_product_order.qty_invoiced, 0.0, "The qty invoiced should be zero as, with the refund, the quantities cancel each other")
+        self.assertEqual(len(self.pol_product_order.invoice_lines), 2, "The product line is invoiced, so it should be linked to 2 invoice lines (invoice and refund)")
+        self.assertEqual(pol_downpayment.qty_to_invoice, -1.0, "As the downpayment was invoiced separately, it will still have to be deducted from the total invoice (hence -1.0), after the refund.")
+        self.assertEqual(pol_downpayment.qty_invoiced, 1.0, "The qty to invoice should be 1 as, with the refund, the products are not invoiced anymore, but the downpayment still is")
+        self.assertEqual(len(pol_downpayment.invoice_lines), 3, "The product line is invoiced, so it should be linked to 3 invoice lines (downpayment invoice, partial invoice and refund)")
+
+    def test_tax_and_account_breakdown(self):
```

Comment:
This test could use a description. At first glance it looks like it's the same as `test_downpayment_percentage_tax_icl`